File: dart/day5_part1_2020.dart

import 'dart:io';

void main() {
  List<String> boardingPasses = File('input.txt').readAsLinesSync();
  
  int maxSeatID = 0;
  
  for (String boardingPass in boardingPasses) {
    int row = int.parse(boardingPass.substring(0, 7).replaceAll('F', '0').replaceAll('B', '1'), radix: 2);
    int column = int.parse(boardingPass.substring(7).replaceAll('L', '0').replaceAll('R', '1'), radix: 2);
    
    int seatID = row * 8 + column;
    
    if (seatID > maxSeatID) {
      maxSeatID = seatID;
    }
  }
  
  print(maxSeatID);
}

File: swift/day5_part1_2020.swift

import Foundation

let input = try String(contentsOfFile: "input.txt")
let boardingPasses = input.components(separatedBy: .newlines)

var maxSeatID = 0

for boardingPass in boardingPasses {
    var rowRange = 0...127
    var colRange = 0...7
    
    for char in boardingPass {
        if char == "F" {
            rowRange = rowRange.lowerBound...(rowRange.lowerBound + rowRange.count/2 - 1)
        } else if char == "B" {
            rowRange = (rowRange.lowerBound + rowRange.count/2)...rowRange.upperBound
        } else if char == "L" {
            colRange = colRange.lowerBound...(colRange.lowerBound + colRange.count/2 - 1)
        } else if char == "R" {
            colRange = (colRange.lowerBound + colRange.count/2)...colRange.upperBound
        }
    }
    
    let seatID = rowRange.lowerBound * 8 + colRange.lowerBound
    maxSeatID = max(maxSeatID, seatID)
}

print(maxSeatID)
